File: apps/backend/src/backend/schemas/track.py
from datetime import datetime

from pydantic import BaseModel, ConfigDict, Field, computed_field

from backend.models.track import AttachmentType


class TrackCreate(BaseModel):
    title: str = Field(..., max_length=255)
    description: str | None = None
    is_public: bool = False
    tags: str | None = Field(None, max_length=500)


class TrackUpdate(BaseModel):
    title: str | None = Field(None, max_length=255)
    description: str | None = None
    is_public: bool | None = None
    tags: str | None = Field(None, max_length=500)


class TrackResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    slug: str
    title: str
    description: str | None
    original_filename: str
    file_size: int
    mime_type: str
    duration_seconds: float | None
    user_id: int | None
    is_public: bool
    tags: str | None
    processing_status: str
    converted_path: str | None
    created_at: datetime
    updated_at: datetime

    @computed_field
    @property
    def stream_url(self) -> str:
        return f"/tracks/{self.id}/stream"


class AttachmentResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    track_id: int
    type: AttachmentType
    content: str | None
    path: str | None
    original_filename: str | None
    caption: str | None
    created_at: datetime
    updated_at: datetime


class TrackDetailResponse(TrackResponse):
    attachments: list[AttachmentResponse] = []


class TrackListResponse(BaseModel):
    items: list[TrackResponse]
    total: int
    page: int
    page_size: int


class AttachmentCreate(BaseModel):
    type: AttachmentType
    content: str | None = None
    caption: str | None = Field(None, max_length=500)


class AttachmentUpdate(BaseModel):
    content: str | None = None
    caption: str | None = Field(None, max_length=500)

File: apps/frontend/src/lib/types/track.ts
export enum AttachmentType {
	NOTE = 'note',
	IMAGE = 'image',
	VIDEO = 'video'
}

export interface TrackResponse {
	id: number;
	slug: string;
	title: string;
	description: string | null;
	original_filename: string;
	file_size: number;
	mime_type: string;
	duration_seconds: number | null;
	user_id: number | null;
	is_public: boolean;
	tags: string | null;
	processing_status: string;
	converted_path: string | null;
	stream_url: string;
	created_at: string;
	updated_at: string;
}

export interface AttachmentResponse {
	id: number;
	track_id: number;
	type: AttachmentType;
	content: string | null;
	path: string | null;
	original_filename: string | null;
	caption: string | null;
	position: number;
	processed_path: string | null;
	processing_status: string;
	file_url: string | null;
	created_at: string;
	updated_at: string;
}

export interface TrackDetailResponse extends TrackResponse {
	attachments: AttachmentResponse[];
}

export interface TrackListResponse {
	items: TrackResponse[];
	total: number;
	page: number;
	page_size: number;
}

export interface TrackCreate {
	title: string;
	description?: string | null;
	is_public?: boolean;
	tags?: string | null;
}

export interface TrackUpdate {
	title?: string | null;
	description?: string | null;
	is_public?: boolean | null;
	tags?: string | null;
}

export interface AttachmentCreate {
	type: AttachmentType;
	content?: string | null;
	caption?: string | null;
}

export interface AttachmentUpdate {
	content?: string | null;
	caption?: string | null;
}

export interface TrackListParams {
	user_id?: number;
	tag?: string;
	page?: number;
	page_size?: number;
}
